plugins/modules/certificate.py

```diff
@@ -0,0 +1,151 @@
+#
+# GNU General Public License v3.0+ (see COPYING or https://www.gnu.org/licenses/gpl-3.0.txt)
+
+from __future__ import absolute_import, division, print_function
+
+__metaclass__ = type
+
+DOCUMENTATION = r"""
+module: certificate
+
+author:
+  - Domen Dobnikar (@domen_dobnikar)
+short_description: Handles cluster SSL certificates.
+description:
+  - Can upload and verify SSL certificates.
+version_added: 1.1.0
+extends_documentation_fragment:
+  - scale_computing.hypercore.cluster_instance
+seealso: []
+options:
+  certificate:
+    description:
+      - File containing the X.509 PEM-encode certificate.
+    type: str
+    required: true
+  private_key:
+    description:
+      - File containing the RSA PEM-encoded private key.
+    type: str
+    required: true
+notes:
+  - C(check_mode) is not supported.
+"""
+
+EXAMPLES = r"""
+- name: Upload new certificate
+  scale_computing.hypercore.certificate:
+    private_key: "{{ lookup('file', 'private_key.key') }}"
+    certificate: "{{ lookup('file', 'scale_cert.cer') }}"
+"""
+
+RETURN = r"""
+record:
+  description:
+    - Certificate record.
+  returned: success
+  type: dict
+  sample:
+    private_key: this_is_the_private_key
+    certificate: this_is_the_certificate
+"""
+
+from ansible.module_utils.basic import AnsibleModule
+
+from ..module_utils import arguments, errors
+from ..module_utils.utils import is_changed
+from ..module_utils.client import Client
+from ..module_utils.rest_client import RestClient
+from ..module_utils.typed_classes import (
+    TypedDiff,
+    TypedTaskTag,
+)
+from ..module_utils.task_tag import TaskTag
+
+from typing import Union, Tuple
+import ssl
+from time import sleep
+
+
+def get_certificate(module: AnsibleModule) -> str:
+    host = (
+        module.params["cluster_instance"]["host"]
+        .replace("https://", "")
+        .replace("http://", "")
+    )
+    cert = ssl.get_server_certificate((host, 443))
+    return cert
+
+
+def upload_cert(module: AnsibleModule, rest_client: RestClient) -> TypedTaskTag:
+    payload = dict(
+        certificate=module.params["certificate"],
+        privateKey=module.params["private_key"],
+    )
+    response = rest_client.create_record("/rest/v1/Certificate", payload, False)
+    return response
+
+
+def ensure_present(
+    module: AnsibleModule,
+    rest_client: RestClient,
+) -> Tuple[bool, Union[str, None], TypedDiff]:
+    before = get_certificate(module)
+    if before.replace("\n", "") == module.params["certificate"].replace("\n", ""):
+        return False, None, dict(before=None, after=None)
+    else:
+        task = upload_cert(module, rest_client)
+        max_retries = 10
+        for ii in range(max_retries):
+            try:
+                TaskTag.wait_task(rest_client, task)
+                break
+            except (
+                errors.ScaleComputingError
+            ) as ex:  # ConnectionRefusedError not working
+                if str(ex) == "[Errno 111] Connection refused":
```

Comment:
Similar problem as with oidc module.
When we upload a certificate the cluster loses connection.

Reply:
You should replace `raise errors.ScaleComputingError(ex)` with just `raise` to re-raise exactly the same exception as the one you intercepted.

The logic to detect exact exception from message inside exception is a bit wrong - what if message for errno code is formatted differently (maybe even due to local language). Can you create issue to refactor this anti-pattern? 
We have is also in `Cluster.shutdown()` - https://github.com/ScaleComputing/HyperCoreAnsibleCollection/blob/main/plugins/module_utils/cluster.py#L77-L101. And maybe on more places.

The problematic line of code (why we do not get something like ConnectionRefusedError) could be in `Client._request()` https://github.com/ScaleComputing/HyperCoreAnsibleCollection/blob/main/plugins/module_utils/client.py#L118-L119

```
        except URLError as e:
            raise ScaleComputingError(e.reason)
```

Reply:
Hmm yes good point I forgot about formatting.
Will create a github issue to look into this and refactor code.

Reply:
I changed code within `client.py` for `ConnectionRefused `and `ConnectionReset`. I also created a github issue to check every other module. 
https://github.com/ScaleComputing/HyperCoreAnsibleCollection/issues/137